csrc/scheduler/reduction.cpp

```diff
@@ -507,19 +507,224 @@ std::shared_ptr<ReductionParams> innerReductionHeuristic(
+
+  // Step-1, set iteration dim
+  // (1) start with bdimx = 8, gidim = 1, iter_unroll = 1
+  hp.bdimx = std::min(8L, hp.iDimAvail());
+  hp.gidim = 1;
+  hp.iter_unroll_factor = 1;
+
+  // (2) increase iter_unroll to its maximum following two rules:
+  // (2.1) ensure divisible split
+  // (2.2) leave enough blocks to saturate the device.
+  // Test on H100 shows the smallest iteration dim we can fully vectorize is
+  // 4800, which corresponds to 75 or 56.8 % of SMs. Here 50% is used.
+  const int64_t min_blocks_fully_vectorize = sm_count / 2;
+  int64_t max_iter_unroll = vectorize_factor;
+  // only allow this adjustment when there is only one output tensor
+  // or it's a small reduction where block reduction is enforced.
+  // When has multiple outputs, reduce vectorization leads to regression,
+  // see cpp benchmark of gelu backward which has 2 outputs ( does't seem like a
+  // common usage since thunder.jit generated fusion has only 1).
+  bool small_reduction = total_reduction_numel <= 1024;
+  if (n_tensor_outputs == 1 || small_reduction) {
+    max_iter_unroll = std::min(
+        max_iter_unroll,
+        scheduler_utils::lastPow2(scheduler_utils::safeDiv(
+            total_iteration_numel, hp.bdimx * min_blocks_fully_vectorize)));
+  }
+  while (hp.iDimAvail() > 1 && hp.iDimAvail() % 2 == 0 &&
+         hp.iter_unroll_factor * 2 <= max_iter_unroll) {
+    hp.iter_unroll_factor *= 2;
+  }
+
+  // (3) increase gidim to SM count, ensures enough blocks to saturate the
+  // device.
+  hp.gidim = std::min(
+      ceilDiv(total_iteration_numel, hp.bdimx * hp.iter_unroll_factor),
+      sm_count);
+
+  // (4) increase bdimx to its maximum
+  hp.bdimx = ceilDiv(total_iteration_numel, hp.gidim * hp.iter_unroll_factor);
+  hp.bdimx = scheduler_utils::roundUpPow2(hp.bdimx);
+  hp.bdimx = std::min(hp.bdimx, max_threads_per_block);
+
+  // (5) re-calculate gidim after bdimx to fix round up differences. Also
+  // handles extreme cases where iter dim is larger than iter_unroll_factor x
+  // max_threads_per_block x sm count
+  hp.gidim = ceilDiv(total_iteration_numel, hp.bdimx * hp.iter_unroll_factor);
+
+  // Step-2, set Reduction dim
+  // (1) reduction unroll takes what is left by iter unroll
+  hp.redu_unroll_factor = std::min(
+      hp.rDimAvail(),
+      scheduler_utils::safeDiv(max_unroll, hp.iter_unroll_factor));
+
+  // (2) bdimy takes what is left by bdimx.
+  hp.bdimy = std::min(hp.rDimAvail(), max_threads_per_block / hp.bdimx);
+
+  // Step-3, final check
+  // (1) revisit bdimx just in case bdimy doesn't take all the left threads
+  int64_t sm_count_pow2 = scheduler_utils::lastPow2(sm_count);
+  while (hp.bdimy * hp.bdimx * 2 <= max_threads_per_block &&
+         hp.gidim / 2 >= sm_count_pow2) {
+    hp.bdimx *= 2;
+    hp.gidim /= 2;
+  }
+
+  // (2) Good heuristic if we have enough blocks to saturate the device or it's
+  // a small reduction. This cut-off is empirical based on H100.
+  const float min_sm_efficiency = 0.9f;
```

Comment:
see the [doc](https://docs.google.com/document/d/12yBhqBHlZFKyxjgXqHjXxOfzZjsjXySHvwR0_8LsfZg/edit#heading=h.rpnkwq2ts553) for why 0.9 is used.